File: KEX/KVG/KVGStroke.m
//
//  KVGStroke.m
//  KEX
//
//  Created by Daniel Schlaug on 3/2/14.
//

#import "KVGStroke.h"
#import "RXMLElement.h"
#import "PocketSVG.h"

@interface KVGStroke ()

@property (strong, nonatomic, readwrite) UIBezierPath *path;
@property (readwrite) unichar type;
@property (readwrite) NSInteger strokeOrder;

@end

@implementation KVGStroke

// Designated initializer
- (id)initFromRXML:(RXMLElement *)pathElement
{
    self = [self init];
    
    //Parse stroke order
    NSString *idAttribute = [pathElement attribute:@"id"];
    NSRegularExpression *regExp = [NSRegularExpression regularExpressionWithPattern:@".*-s([1-9][0-9]?)(-.*|$)" options:0 error:nil];
    NSString *strokeOrderString = [regExp stringByReplacingMatchesInString:idAttribute options:0 range:(NSRange){0, [idAttribute length]} withTemplate:@"$1"];
    self.strokeOrder = [strokeOrderString integerValue];
    NSAssert(self.strokeOrder != 0, @"Couldn't parse stroke order: %@", idAttribute);
    
    //Parse stroke path
    CGPathRef cgPath = [PocketSVG pathFromDAttribute:[pathElement attribute:@"d"]];
    self.path = [UIBezierPath bezierPathWithCGPath:cgPath];
    
    //Parse stroke type
    self.type = [[pathElement attribute:@"type"] characterAtIndex:0];
    
    return self;
}

@end

File: KEX/KVG/KVGStroke.swift
//
//  KVGStroke.swift
//  KEX
//
//  Created by Daniel Schlaug on 6/25/14.
//  Copyright (c) 2014 Daniel Schlaug. All rights reserved.
//

import Foundation
import CoreGraphics
import DanielsKit

class KVGStroke: NSObject {
    
    let path: UIBezierPath
    let type: Character
    let strokeOrder: Int = 0
    
    init(RXML pathElement: RXMLElement) {
        let idAttribute = pathElement.attribute("id")
        let strokeOrderRegExp = ~/".*-s([1-9][0-9]?)(-.*|$)"
        let strokeOrderString: NSString = strokeOrderRegExp.stringByReplacingMatchesInString(idAttribute, options: nil, range: idAttribute.fullRange, withTemplate: "$1")
        let strokeOrder = strokeOrderString.integerValue
        self.strokeOrder = strokeOrder
        assert(strokeOrder != 0, "Could not parse stroke order.")
        
        
        let aCGPath: CGPath = PocketSVG.pathFromDAttribute(pathElement.attribute("d")).takeRetainedValue() //WARNING not sure that this should be retained
        self.path = UIBezierPath(CGPath: aCGPath)
        
        self.type = pathElement.attribute("type")[0]
    }
}